### PR TITLE
tasty-1.4.2.3: Eliminate dependency on mtl

### PR DESCRIPTION
**Pull request by Bodigrim**
As suggested at https://github.com/UnkindPartition/tasty/pull/334#issuecomment-1120489902
This will allow `mtl` to have a test suite itself, without a circular dependency.

### CODE REVIEW

File: core/Test/Tasty/Run.hs [hunk 1]
```diff
@@ -279,7 +280,7 @@ createTestActions opts0 tree = do
       let
         ini = Initializer doInit initVar
         fin = Finalizer doRelease initVar finishVar
-        tests' = map (first $ local $ (Seq.|> ini) *** (fin Seq.<|)) tests
+        tests' = map (first (. ((Seq.|> ini) *** (fin Seq.<|)))) tests
```
Comment (andreasabel) [inline core/Test/Tasty/Run.hs:283]: Please make this more pointfull.  This is impossible to comprehend.

File: core/Test/Tasty/Patterns/Eval.hs [hunk 2]
```diff
@@ -49,8 +49,8 @@ asN v =
     VS True s ->
       case parseN s of
         Just n -> return n
-        Nothing -> throwError $ "Not a number: " ++ show s
-    VS False s -> throwError $ "String is not numeric: " ++ show s
+        Nothing -> lift $ Left $ "Not a number: " ++ show s
```
Comment (andreasabel) [inline core/Test/Tasty/Patterns/Eval.hs:52]: Please define a `throwError` for monad `M` and use this here.

File: core/Test/Tasty/Run.hs [hunk 3]
```diff
@@ -244,9 +245,9 @@ createTestActions opts0 tree = do
           { foldSingle = runSingleTest
           , foldResource = addInitAndRelease
           , foldGroup = \_opts name (Traversal a) ->
-              Traversal $ local (first (Seq.|> name)) a
+              Traversal $ (mapWriterT . local) (first (Seq.|> name)) a
```
Comment (andreasabel) [inline core/Test/Tasty/Run.hs:248]: A more pointfull style please, to preserve code readability.

**Review by andreasabel — CHANGES_REQUESTED** [review]

**Comment by Bodigrim** [comment]
@andreasabel done.

**Review by andreasabel — APPROVED** [review]

**Comment by andreasabel** [comment]
Candidate at: https://hackage.haskell.org/package/tasty-1.4.2.3/candidate

**Pull request merged**